core/src/main/java/ai/timefold/solver/core/impl/domain/variable/ListVariableState.java

```diff
@@ -0,0 +1,291 @@
+            externalizedInverseProcessor.changeElement(scoreDirector, entity, element);
+        }
+        // Next and previous still might have changed, even if the index and entity did not.
+        // Those are based on what happened elsewhere in the list.
+        if (externalizedPreviousElementProcessor != null) {
+            externalizedPreviousElementProcessor.setElement(scoreDirector, elements, element, index);
+        }
+        if (externalizedNextElementProcessor != null) {
+            externalizedNextElementProcessor.setElement(scoreDirector, elements, element, index);
+        }
+        return difference.anythingChanged;
+    }
+
+    private ChangeType processElementLocation(Object entity, Object element, int index) {
+        if (requiresLocationMap) { // Update the location and figure out if it is different from previous.
+            var newLocation = ElementLocation.of(entity, index);
+            var oldLocation = elementLocationMap.put(element, newLocation);
+            if (oldLocation == null) {
+                unassignedCount--;
+                return ChangeType.BOTH;
+            }
+            return compareLocations(entity, oldLocation.entity(), index, oldLocation.index());
+        } else { // Read the location and figure out if it is different from previous.
+            var oldEntity = getInverseSingleton(element);
+            if (oldEntity == null) {
+                unassignedCount--;
+                return ChangeType.BOTH;
+            }
+            var oldIndex = getIndex(element);
+            if (oldIndex == null) { // Technically impossible, but we handle it anyway.
+                return ChangeType.BOTH;
+            }
+            return compareLocations(entity, oldEntity, index, oldIndex);
+        }
+    }
+
+    private static ChangeType compareLocations(Object entity, Object otherEntity, int index, int otherIndex) {
+        if (entity != otherEntity) {
+            return ChangeType.BOTH; // Entity changed, so index changed too.
+        } else if (index != otherIndex) {
+            return ChangeType.INDEX;
+        } else {
+            return ChangeType.NEITHER;
+        }
+    }
+
+    private enum ChangeType {
+
+        BOTH(true, true),
+        INDEX(false, true),
+        NEITHER(false, false);
+
+        final boolean anythingChanged;
+        final boolean entityChanged;
+        final boolean indexChanged;
+
+        ChangeType(boolean entityChanged, boolean indexChanged) {
+            this.anythingChanged = entityChanged || indexChanged;
+            this.entityChanged = entityChanged;
+            this.indexChanged = indexChanged;
+        }
+
+    }
+
+    public ElementLocation getLocationInList(Object planningValue) {
+        if (!canUseExternalizedLocation) {
+            return Objects.requireNonNullElse(elementLocationMap.get(planningValue), ElementLocation.unassigned());
+        } else {
+            var inverse = getInverseSingleton(planningValue);
```

Comment:
`getInverseSingleton` and `getIndex` might do a map lookup on the same value. Might be worth it to do a single map lookup here if requiresLocationMap is true.